include/revng-c/RestructureCFG/CycleEquivalenceAnalysis.h

```diff
@@ -0,0 +1,113 @@
+#pragma once
+
+//
+// Copyright rev.ng Labs Srl. See LICENSE.md for details.
+//
+
+#include <map>
+
+#include "llvm/IR/CFG.h"
+#include "llvm/Pass.h"
+
+#include "revng/ADT/GenericGraph.h"
+
+#include "CycleEquivalenceClass.h"
+
+// `struct` used to represent the unique incremental `Index` that is
+// associated with each undirected edge in the `GenericGraph`
+struct EdgeIndex {
+private:
+  size_t Index = 0;
+
+public:
+  EdgeIndex(size_t Index) { this->Index = Index; }
+  size_t getValue() const { return Index; }
+  bool operator<(const EdgeIndex &Rhs) const { return Index < Rhs.getValue(); }
+  bool operator==(const EdgeIndex &Rhs) const {
+    return Index == Rhs.getValue();
+  }
+};
+
+template<class GraphT, class GT = llvm::GraphTraits<GraphT>>
+class CycleEquivalenceAnalysis {
+public:
+  using NodeT = typename GT::NodeRef;
+
+private:
+  struct BlockBidirectionalNode {
+    BlockBidirectionalNode(NodeT Block) : Block(Block) {}
+    NodeT Block;
+    NodeT getBlock() { return Block; }
+    llvm::StringRef getName() const { return Block->getName(); }
+  };
+
+public:
+  using EdgeLabel = EdgeIndex;
+  using BlockNode = BidirectionalNode<BlockBidirectionalNode, EdgeLabel>;
+  using BlockGraph = GenericGraph<BlockNode>;
+  using EdgeDescriptor = revng::detail::EdgeDescriptor<BlockNode *>;
+  using EdgesVectorType = llvm::SmallVector<EdgeDescriptor>;
+  using BracketDescriptor = std::pair<EdgeDescriptor, size_t>;
+
+private:
+  llvm::SmallVector<CycleEquivalenceClass<NodeT>, 1> CycleEquivalenceClasses;
+
+  std::map<BracketDescriptor, size_t> BracketToClass;
+  std::map<size_t, BracketDescriptor> ClassToBracketDescriptor;
```

Comment:
Mmm wait, I think you misunderstood some detail here.

This map is just used to keep a map between the n-th equivalence class, and the `BracketDescriptor` that gave origin to it. E.g., say we at some point create a new equivalence class, which is the first one, with some edges in it, described by the bracket `(a,1)`, we note down that the 0-th equivalence class is originated by such `BracketDescriptor`.

This information is used only when printing debug information, but I would be very sad removing it, since this information is very useful to debug the internal functioning of  `CycleEquivalenceAnalysis` in case of bugs.

Reply:
Yeah, you understood it correctly, at the moment this map is used just to keep track of the original extensive name of the `bracket` which gave origin to a specific equivalence class. The algorithm per se, doesn't need this mapping, but it is useful for debug prints in case of bug hunting. Added a comment explaining this here https://github.com/revng/revng-c/blob/ce352f2912e639e1ef307bb1d5b131db48444878/include/revng-c/RestructureCFG/CycleEquivalenceAnalysis.h#L201